rust/private/toolchain_utils.bzl

```diff
@@ -1,5 +1,171 @@
+
+    Returns:
+        depset[File]: A depset of the generated symlink files
+    """
+    tree_files = []
+    for file in target.files.to_list():
+        # Parse the path to the file relative to the workspace root so a
+        # symlink matching this path can be created within the sysroot.
+
+        # The code blow attempts to parse any workspace names out of the
+        # path. For local targets, this code is a noop.
+        if target.label.workspace_root:
+            file_path = file.path.split(target.label.workspace_root, 1)[-1]
+        else:
+            file_path = file.path
+
+        symlink = ctx.actions.declare_file("{}/{}".format(name, file_path.lstrip("/")))
+
+        ctx.actions.symlink(
+            output = symlink,
+            target_file = file,
+        )
+
+        tree_files.append(symlink)
+
+    return depset(tree_files)
+
+def _symlink_sysroot_bin(ctx, name, directory, target):
+    """Crete a symlink to a target file.
+
+    Args:
+        ctx (ctx): The rule's context object
+        name (str): A common name for the output directory
+        directory (str): The directory under `name` to put the file in
+        target (File): A File object to symlink to
+
+    Returns:
+        File: A newly generated symlink file
+    """
+    symlink = ctx.actions.declare_file("{}/{}/{}".format(
+        name,
+        directory,
+        target.basename,
+    ))
+
+    ctx.actions.symlink(
+        output = symlink,
+        target_file = target,
+        is_executable = True,
+    )
+
+    return symlink
+
+def generate_sysroot(
+        ctx,
+        rustc,
+        rustdoc,
+        rustc_lib,
+        cargo = None,
+        clippy = None,
+        llvm_tools = None,
+        rust_std = None,
+        rustfmt = None):
+    """Generate a rust sysroot from an exec and target toolchain
+
+    Args:
+        ctx (ctx): A context object from a `rust_toolchain` rule.
+        rustc (File): The path to a `rustc` executable.
+        rustdoc (File): The path to a `rustdoc` executable.
+        rustc_lib (Target): A collection of Files containing dependencies of `rustc`.
+        cargo (File, optional): The path to a `cargo` executable.
+        clippy (File, optional): The path to a `clippy-driver` executable.
+        llvm_tools (Target, optional): A collection of llvm tools used by `rustc`.
+        rust_std (Target, optional): A collection of Files containing Rust standard library components.
+        rustfmt (File, optional): The path to a `rustfmt` executable.
+
+    Returns:
+        struct: A struct of generated files representing the new sysroot
+    """
+    name = ctx.label.name
+
+    # Define runfiles
+    direct_files = []
+    transitive_file_sets = []
+
+    # Rustc
+    sysroot_rustc = _symlink_sysroot_bin(ctx, name, "bin", rustc)
+    direct_files.extend([sysroot_rustc, rustc])
+
+    # Rustc dependencies
+    sysroot_rustc_lib = _symlink_sysroot_tree(ctx, name, rustc_lib) if rustc_lib else None
```

Comment:
Nit: I find the following format more readable, and more common

```
sysroot_rustc_lib = None
if rustc_lib:
	sysroot_rustc_lib = _symlink_sysroot_tree(ctx, name, rustc_lib)
        transitive_file_sets.extend([sysroot_rustc_lib, rustc_lib.files])
```

Reply:
Always nervous about the number of lines changed in a PR. I feel more people look at that number and immediately ignore it after it hits their threshold. But since you asked I'm happy to update this 😄